osrf/vrx
# Scoring plugin part 4 - Terminate simulation

This pull requests terminates the simulation and the ROS infrastructure when needed. We trigger the shutdown when the SDF parameter `<per_plugin_exit_on_completion>`  or the `VRX_EXIT_ON_COMPLETION` environment variable are set to `true`. Otherwise, the simulation should keep running even if we transition to `finished` state.

### How to test it?

1. I recommend changing the `running_state_duration` parameter to `10` for rapid testing.

    <details><summary>patch</summary>

    ````diff
    diff --git a/vrx_ign/worlds/sydney_regatta.sdf b/vrx_ign/worlds/sydney_regatta.sdf
    index 44b13c8c..694e73c0 100644
    --- a/vrx_ign/worlds/sydney_regatta.sdf
    +++ b/vrx_ign/worlds/sydney_regatta.sdf
    @@ -678,7 +678,7 @@
           <task_info_topic>/vrx/task/info</task_info_topic>
           <initial_state_duration>10</initial_state_duration>
           <ready_state_duration>10</ready_state_duration>
    -      <running_state_duration>300</running_state_duration>
    +      <running_state_duration>10</running_state_duration>
           <per_plugin_exit_on_completion>false</per_plugin_exit_on_completion>
           <release_joints>
             <joint>

2. Launch the simulation:
```
ros2 launch vrx_ros competition_local.launch.py ign_args:="-v 4 -r sydney_regatta.sdf"
```

And verify that after `30` seconds of simulation time, the simulation keeps going and doesn't terminate.

3. Launch the simulation with `VRX_EXIT_ON_COMPLETITION` set to true:

```
VRX_EXIT_ON_COMPLETION=true ros2 launch vrx_ros competition_local.launch.py ign_args:="-v 4 -r sydney_regatta.sdf"
```

This should terminate all the ROS processes and the simulation after `30` seconds.

4. Set the `<per_plugin_exit_on_completition>` to true and launch the simulation:

    <details><summary>patch</summary>

    ````diff
    diff --git a/vrx_ign/worlds/sydney_regatta.sdf b/vrx_ign/worlds/sydney_regatta.sdf
    index 44b13c8c..6dc23b22 100644
    --- a/vrx_ign/worlds/sydney_regatta.sdf
    +++ b/vrx_ign/worlds/sydney_regatta.sdf
    @@ -679,7 +679,7 @@
           <initial_state_duration>10</initial_state_duration>
           <ready_state_duration>10</ready_state_duration>
           <running_state_duration>300</running_state_duration>
    -      <per_plugin_exit_on_completion>false</per_plugin_exit_on_completion>
    +      <per_plugin_exit_on_completion>true</per_plugin_exit_on_completion>
           <release_joints>
             <joint>
               <name>wamv_external_pivot_joint</name>

```
ros2 launch vrx_ros competition_local.launch.py ign_args:="-v 4 -r sydney_regatta.sdf"
```

This should terminate all the ROS processes and the simulation after `30` seconds.
